### DISCUSSION TITLE
File wildcard/glob patterns in Windows eza

### DISCUSSION THREAD

**Discussion by TxAggie87** [eza-community/eza#621] [Q&A, Howto, Help] [answered]
I first used exa and have now been using eza with various Linux installs for years now.  It's a great tool.  I have just tried installing it on a windows machine for the first time and the winget installer fails (cannot extract the downloaded zip file).  But I was able to manually extract and install the executable.  I can run the eza command and it looks great and uses the same options as the Linux install. :-)

However, if I try to include any filename wildcard/glob pattern, it always reports: "The filename, directory name, or volume label syntax is incorrect. (os error 123)".  If I exactly specify a complete filename it works and shows that one file.  If I don't specify a filename, it works and shows all files/directories.  I cannot figure out why I cannot get any file globs to work.  I have tried both standard windows pattterns (w*.*) as well as Linux style patterns.

I know windows isn't the real focus for this application, but does anyone have an ideas as to what I might be doing wrong?

Thx.

**Comment by TxAggie87** [comment, marked as answer]
OK, I apologize for my brain freeze.  When I write code in Windows, I forget that the .net file functions typically do the glob expansion and that in the rest of the world it tends to be the shell handling this need.  

I created a small .cmd file named glob.cmd:
```
@echo off
set expanded_list=
for /f "tokens=*" %%F in ('dir /b /a:-d "%~1"') do call set expanded_list=%%expanded_list%% "%%F"

echo expanded_list is:
echo %expanded_list%
```
If I run my glob.cmd as: glob.cmd w*.*
and then run: eza -lhag %expanded_list%
it works as expected.  It's not a complete solutions and it doesn't handle recursive globbing, but that can also be resolved.  So, because of limitations in Windows, this is not as convenient as it is in Linux.

If anyone has better solution, I'd be interested....

Thx.

**Comment by eggbean** [comment]
This is a severe limitation for eza on Windows.

Yes, apparently that while on Linux globbing is handled by the shell, on Windows it's up to the program to handle it. And...

> MSVC has an [optional way to compile a program](https://learn.microsoft.com/en-us/cpp/c-language/expanding-wildcard-arguments?view=msvc-170), that makes the C runtime expand wildcards in the command line before the program actually receives the command line. It can be used to make a program behave similar to as if it were launched from a Linux shell. Typically, when programs are compiled using that optional mode it creates confusion and quirky edge case problems, but the mode exists and is possible for an app author to use if they wish.

https://github.com/chrisant996/clink/discussions/548#discussioncomment-8137917

I don't know anything about compiling on Windows, but I want to do what I can to solve this problem. Quite a few of the ported Unix coreutils don't have this problem (some do), so it seems that this is something that can be fixed on eza for Windows.

**Reply by daviessm** [reply]
@eggbean if you can find any Rust programs that use the above trick on Windows, better still if you can find the config that does it, I can create an `eza` Windows build with it for you to try.

**Reply by eggbean** [reply]
@daviessm I use the scoop package for [uutils-coreutils](https://github.com/uutils/coreutils) which is a Rust rewrite of the GNU coreutils. As far as I have tested, none of these have this problem, including `ls.exe`.

**Reply by daviessm** [reply]
I think https://github.com/uutils/coreutils/pull/1281 is the change that fixed it in `uutils`, I'll take a closer look when I get a moment.

**Reply by daviessm** [reply]
This will almost certainly be changed by the clap PR https://github.com/eza-community/eza/pull/741 as well so any changes will have to go in after that. There's a possibility that PR also accidentally fixes this issue too...

**Reply by eggbean** [reply]
Ha, that would be cool.

**Reply by daviessm** [reply]
Can you try this branch and see if it works for you? https://github.com/daviessm/eza/tree/clap

Edit: @TxAggie87 you as well!

**Reply by eggbean** [reply]
Seems to work great 🎉

**Reply by eggbean** [reply]
What's happened? It doesn't seem to work anymore?

**Reply by daviessm** [reply]
I haven't touched that branch or put in a pull request to `main` yet because I'm waiting for other changes to be merged first.

**Reply by eggbean** [reply]
Oh, I thought it had already been merged.

**Reply by per11235813** [reply]
PR #741 has been merged to the branch v1.0-development. The problem with globs on windows persists on this branch.

@daviessm fork at https://github.com/daviessm/eza/tree/clap does work well.

I hope @daviessm changes will be added to eza.

**Reply by daviessm** [reply]
I'll do the PR once #741 has been merged.

**Reply by eggbean** [reply]
What does that mean, as it says that it has been merged on the page?

**Reply by daviessm** [reply]
It's been merged into a separate branch for final testing, it's not in the main branch yet.

**Reply by eggbean** [reply]
> I'll do the PR once #741 has been merged.

Have you made the PR yet? I'm still using the old test build so that I can use wildcards on Windows.

**Comment by mattdkerr** [comment]
I'd love to have [this ](https://github.com/daviessm/eza/tree/clap)merged-- it's still an issue on Windows. I can switch to Git Bash but it's still non-functional
```
eza -R *.proto
zsh: no matches found: *.proto
```
```
eza -R **/*.proto
zsh: argument list too long: eza
```
I built eza from `cargo install eza` today.
